File: src/ados/sdk/drivers/gimbal.py
"""Gimbal driver base class.

A gimbal driver moves a stabilised mount to a commanded attitude or
rate, and reports back the current pointing state. Vendor-specific
serial mounts, SBGC-family controllers, and MAVLink mount protocols
all share this interface.
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Any, AsyncIterator


@dataclass(frozen=True)
class GimbalCandidate:
    """A gimbal a driver claims it can open."""

    driver_id: str
    device_id: str
    label: str
    bus: str
    vid_pid: tuple[int, int] | None = None
    metadata: dict[str, Any] | None = None


@dataclass(frozen=True)
class GimbalCapabilities:
    """Static capabilities of an open gimbal session.

    Pitch, yaw, and roll limits are degrees from neutral. ``None`` on a
    rate limit means the axis is position-controlled only.
    """

    has_pitch: bool
    has_yaw: bool
    has_roll: bool
    pitch_min_deg: float
    pitch_max_deg: float
    yaw_min_deg: float
    yaw_max_deg: float
    roll_min_deg: float
    roll_max_deg: float
    max_rate_dps: float | None = None
    supports_follow_mode: bool = False
    supports_lock_mode: bool = False


@dataclass(frozen=True)
class GimbalState:
    """A single state sample reported by a gimbal."""

    timestamp_ns: int
    pitch_deg: float
    yaw_deg: float
    roll_deg: float
    pitch_rate_dps: float = 0.0
    yaw_rate_dps: float = 0.0
    roll_rate_dps: float = 0.0
    mode: str = "neutral"
    metadata: dict[str, Any] | None = None


class GimbalSession:
    """Opaque per-open state held by a gimbal driver."""


class GimbalDriver(ABC):
    """Abstract base for gimbal drivers."""

    @abstractmethod
    async def discover(self) -> list[GimbalCandidate]:
        """Scan for gimbals this driver can open."""

    @abstractmethod
    async def open(
        self, candidate: GimbalCandidate, config: dict[str, Any]
    ) -> GimbalSession:
        """Open a session against a gimbal."""

    @abstractmethod
    async def close(self, session: GimbalSession) -> None:
        """Release resources held by a session."""

    @abstractmethod
    def capabilities(self, session: GimbalSession) -> GimbalCapabilities:
        """Return the static capabilities of an open session."""

    @abstractmethod
    async def command_attitude(
        self,
        session: GimbalSession,
        pitch_deg: float,
        yaw_deg: float,
        roll_deg: float = 0.0,
    ) -> None:
        """Drive the gimbal to an absolute pitch, yaw, and roll setpoint.

        Out-of-range setpoints should be clamped against
        :class:`GimbalCapabilities` rather than silently ignored.
        """

    @abstractmethod
    async def command_rate(
        self,
        session: GimbalSession,
        pitch_rate_dps: float,
        yaw_rate_dps: float,
        roll_rate_dps: float = 0.0,
    ) -> None:
        """Drive the gimbal at a commanded angular rate.

        Drivers without rate control should raise :class:`NotImplementedError`
        so the host can fall back to position commands.
        """

    @abstractmethod
    def get_state(self, session: GimbalSession) -> GimbalState:
        """Return the most recent attitude sample."""

    @abstractmethod
    async def state_iterator(
        self, session: GimbalSession
    ) -> AsyncIterator[GimbalState]:
        """Yield attitude samples until the session closes."""

File: crates/ados-sdk/src/drivers/gimbal.rs
//! Gimbal driver trait.
//!
//! Ports `ados.sdk.drivers.gimbal`. A gimbal driver moves a stabilised mount to
//! a commanded attitude or rate and reports the pointing state. Vendor serial
//! mounts, SBGC-family controllers, and MAVLink mount protocols share this
//! interface.

use std::collections::BTreeMap;

use async_trait::async_trait;
use rmpv::Value;

use super::{Bus, DriverResult, SampleStream};

/// A gimbal a driver claims it can open.
#[derive(Debug, Clone, PartialEq)]
pub struct GimbalCandidate {
    pub driver_id: String,
    pub device_id: String,
    pub label: String,
    pub bus: Bus,
    pub vid_pid: Option<(u16, u16)>,
    pub metadata: BTreeMap<String, Value>,
}

/// Static capabilities of an open gimbal session. Limits are degrees from
/// neutral. `max_rate_dps == None` on an axis means it is position-controlled
/// only.
#[derive(Debug, Clone, PartialEq)]
pub struct GimbalCapabilities {
    pub has_pitch: bool,
    pub has_yaw: bool,
    pub has_roll: bool,
    pub pitch_min_deg: f32,
    pub pitch_max_deg: f32,
    pub yaw_min_deg: f32,
    pub yaw_max_deg: f32,
    pub roll_min_deg: f32,
    pub roll_max_deg: f32,
    pub max_rate_dps: Option<f32>,
    pub supports_follow_mode: bool,
    pub supports_lock_mode: bool,
}

/// A single state sample reported by a gimbal.
#[derive(Debug, Clone, PartialEq)]
pub struct GimbalState {
    pub timestamp_ns: i64,
    pub pitch_deg: f32,
    pub yaw_deg: f32,
    pub roll_deg: f32,
    pub pitch_rate_dps: f32,
    pub yaw_rate_dps: f32,
    pub roll_rate_dps: f32,
    pub mode: String,
    pub metadata: BTreeMap<String, Value>,
}

/// Abstract gimbal driver.
#[async_trait]
pub trait GimbalDriver: Send + Sync {
    type Session: Send + Sync;

    /// Scan for gimbals this driver can open.
    async fn discover(&self) -> DriverResult<Vec<GimbalCandidate>>;

    /// Open a session against a gimbal.
    async fn open(
        &self,
        candidate: &GimbalCandidate,
        config: &BTreeMap<String, Value>,
    ) -> DriverResult<Self::Session>;

    /// Release resources held by a session.
    async fn close(&self, session: Self::Session) -> DriverResult<()>;

    /// Return the static capabilities of an open session.
    fn capabilities(&self, session: &Self::Session) -> GimbalCapabilities;

    /// Drive the gimbal to an absolute pitch/yaw/roll setpoint. Out-of-range
    /// setpoints should be clamped against [`GimbalCapabilities`] rather than
    /// silently ignored.
    async fn command_attitude(
        &self,
        session: &Self::Session,
        pitch_deg: f32,
        yaw_deg: f32,
        roll_deg: f32,
    ) -> DriverResult<()>;

    /// Drive the gimbal at a commanded angular rate. Drivers without rate
    /// control return [`super::DriverError::NotSupported`] so the host falls
    /// back to position commands.
    async fn command_rate(
        &self,
        session: &Self::Session,
        pitch_rate_dps: f32,
        yaw_rate_dps: f32,
        roll_rate_dps: f32,
    ) -> DriverResult<()>;

    /// Return the most recent attitude sample.
    fn get_state(&self, session: &Self::Session) -> GimbalState;

    /// Yield attitude samples until the session closes.
    async fn state_iterator(
        &self,
        session: &Self::Session,
    ) -> DriverResult<SampleStream<GimbalState>>;
}
